Error with Menhir 20211230

With the [latest release of menhir](https://gitlab.inria.fr/fpottier/menhir/blob/master/CHANGES.md#anchor-20211230) that requires type information in the source or inference enabled:

> The code back-end now needs type information. This means that either Menhir's type inference mechanism must be enabled (the easiest way of enabling it is to use Menhir via dune and to check that the dune-project file says (using menhir 2.0) or later) or the type of every nonterminal symbol must be explicitly given via a %type declaration.

The menhir backend for BNFC currently fails as follows:

```
❯ cat > Sum.cf
EInt.  Exp ::= Integer;
EPlus. Exp ::= Exp "+" Integer;
/tmp
❯ bnfc -m Sum.cf --ocaml-menhir &&  make
2 rules accepted

writing new file ./AbsSum.ml
writing new file ./LexSum.mll
writing new file ./ParSum.mly
writing new file ./SkelSum.ml
writing new file ./PrintSum.ml
writing new file ./ShowSum.ml
writing new file ./TestSum.ml
writing new file ./BNFC_Util.ml
writing new file ./Makefile
menhir ParSum.mly
File "ParSum.mly", line 14, characters 14-22:
Warning: the token TOK_Char is unused.
File "ParSum.mly", line 15, characters 15-25:
Warning: the token TOK_Double is unused.
File "ParSum.mly", line 13, characters 16-25:
Warning: the token TOK_Ident is unused.
File "ParSum.mly", line 17, characters 16-26:
Warning: the token TOK_String is unused.
Error: the code back-end requires the type of every nonterminal symbol to be
known. Please specify the type of every symbol via %type declarations, or
enable type inference (look up --infer in the manual).
Type inference is automatically enabled when Menhir is used via Dune,
provided the dune-project file says (using menhir 2.0) or later.
The types of the following nonterminal symbols are unknown:
exp
int
make: *** [all] Error 1
/tmp
```

--- a/source/CHANGELOG.md
+++ b/source/CHANGELOG.md
@@ -6,6 +6,7 @@ Unreleased
 * Agda: support position information via `--functor` [#405]
 * C/C++: use `size_t` and `-Wsign-conversion` [#391]
 * C++: repair broken `--line-numbers` [#390], regression in 2.9.2 by [#349]
+* Ocaml: non-terminals in generated parser are now type-annotated [#407]
 
 Contributors: Michał Radwański
 

--- a/source/src/BNFC/Backend/OCaml/CFtoOCamlYacc.hs
+++ b/source/src/BNFC/Backend/OCaml/CFtoOCamlYacc.hs
@@ -15,11 +15,12 @@ module BNFC.Backend.OCaml.CFtoOCamlYacc
         where
 
 import Data.Char
-import Data.Foldable (toList)
+import Data.Foldable ( toList )
+import Data.List     ( intercalate )
 
 import BNFC.CF
-import BNFC.Options ( OCamlParser(..) )
-import BNFC.Utils   ( (+++) )
+import BNFC.Options  ( OCamlParser(..) )
+import BNFC.Utils    ( (+++), mapHead )
 import BNFC.Backend.Common
 import BNFC.Backend.OCaml.OCamlUtil
 
@@ -32,13 +33,13 @@ type MetaVar     = String
 -- The main function, that given a CF
 -- generates a ocamlyacc module.
 cf2ocamlyacc :: OCamlParser -> String -> String -> String -> CF -> String
-cf2ocamlyacc ocamlParser name absName lexName cf
- = unlines
-    [header name absName lexName cf,
-    declarations absName cf,
-    "%%",
-    rules ocamlParser cf
-    ]
+cf2ocamlyacc ocamlParser name absName lexName cf = unlines
+  [ header name absName lexName cf
+  , declarations absName cf
+  , "%%"
+  , ""
+  , rules ocamlParser cf
+  ]
 
 
 header :: String -> String -> String -> CF -> String
@@ -53,17 +54,21 @@ header _ absName _ _ = unlines
 
 declarations :: String -> CF -> String
 declarations absName cf =
-  unlines
+  unlines $ intercalate [""]
     [ tokens (unicodeAndSymbols cf) (asciiKeywords cf)
     , specialTokens cf
     , entryPoints absName cf
+    , map (catTyping . fst)      $ ruleGroups cf
+    , map (catTyping . TokenCat) $ literals cf
     ]
+  where
+    catTyping c = typing absName c (nonterminal c)
 
 -- | Declare keyword and symbol tokens.
 
-tokens :: [String] -> [String] -> String
+tokens :: [String] -> [String] -> [String]
 tokens symbols reswords =
-  unlines $ concat
+  concat
     [ [ unwords $ "%token" : map ("KW_" ++) reswords | hasReserved ]
     , [ "" | hasReserved ]
     , (`map` zip symbols [1::Int ..]) $ \ (s, n) ->
@@ -93,8 +98,8 @@ nonterminal c = map spaceToUnderscore (fixType c)
     where spaceToUnderscore ' ' = '_'
           spaceToUnderscore x = x
 
-specialTokens :: CF -> String
-specialTokens cf = unlines $ concat $
+specialTokens :: CF -> [String]
+specialTokens cf = concat $
   [ [ "%token TOK_EOF" ]
   , [ prToken (ty n)      n | n                 <- specialCatsP  ]
   , [ prToken (posTy pos) n | TokenReg n0 pos _ <- cfgPragmas cf, let n = wpThing n0 ]
@@ -112,15 +117,18 @@ specialTokens cf = unlines $ concat $
     True  -> "<(int * int) * string>"
     False -> "<string>"
 
+entryPoints :: String -> CF -> [String]
+entryPoints absName cf =
+  concat
+    [ [ unwords $ "%start" : map epName eps ]
+    , map (\ c -> typing absName c (epName c)) eps
+    ]
+  where
+    eps = toList $ allEntryPoints cf
 
-entryPoints :: String -> CF -> String
-entryPoints absName cf = unlines $
-    ("%start" +++ unwords (map epName eps))
-    :
-    (map typing eps)
-    where eps = toList $ allEntryPoints cf
-          typing :: Cat -> String
-          typing c = "%type" +++ "<" ++ qualify (normCat c) ++ ">" +++ epName c
+typing :: String -> Cat -> String -> String
+typing absName c s = "%type" +++ "<" ++ qualify (normCat c) ++ ">" +++ s
+    where
           qualify c = if c `elem` [ TokenCat "Integer", TokenCat "Double", TokenCat "Char",
                                     TokenCat "String", ListCat (TokenCat "Integer"),
                                     ListCat (TokenCat "Double"),
@@ -130,18 +138,15 @@ entryPoints absName cf = unlines $
                       else absName ++ "." ++ fixType c
 
 epName :: Cat -> String
-epName c = "p" ++ capitalize (nonterminal c)
-            where capitalize s = case s of
-                    [] -> []
-                    c:cs -> toUpper c : cs
+epName c = "p" ++ mapHead toUpper (nonterminal c)
 
-entryPointRules :: OCamlParser -> CF -> String
+entryPointRules :: OCamlParser -> CF -> [String]
 entryPointRules ocamlParser cf =
-  unlines $ map mkRule $ toList $ allEntryPoints cf
+  map mkRule $ toList $ allEntryPoints cf
   where
   mkRule :: Cat -> String
-  mkRule s = unlines
-    [ epName s ++ " : " ++ nonterminal s ++ " TOK_EOF { $1 }"
+  mkRule cat = unlines
+    [ epName cat ++ " : " ++ nonterminal cat ++ " TOK_EOF { $1 }"
     , concat
       [ "  | error { raise (BNFC_Util.Parse_error ("
         -- Andreas, 2021-09-16, issue #380
@@ -160,22 +165,22 @@ entryPointRules ocamlParser cf =
     ]
 
 rules :: OCamlParser -> CF -> String
-rules ocamlParser cf = unlines [
-    entryPointRules ocamlParser cf,
-    (unlines $ map (prOne . mkOne) (ruleGroups cf)),
-    specialRules cf
+rules ocamlParser cf = unlines $ concat
+    [ entryPointRules ocamlParser cf
+    , map (prOne . mkOne) $ ruleGroups cf
+    , specialRules cf
     ]
-    where
-        mkOne (cat,rules) = (cat, constructRule (terminal cf) rules cat)
-        prOne (_,[]) = [] -- nt has only internal use
-        prOne (nt,((p,a):ls)) =
-          unwords [nt', ":" , p, "{", a, "}", "\n" ++ pr ls] ++ ";\n"
-         where
-           nt' = nonterminal nt
-           pr [] = []
-           pr ((p,a):ls) =
-             unlines [ unwords [ "  |", p, "{", a , "}" ] ] ++ pr ls
-
+  where
+    mkOne (cat,rules) = (cat, constructRule (terminal cf) rules cat)
+    prOne (_  , []  ) = [] -- nt has only internal use
+    prOne (cat, l:ls) = unlines $ concat
+        [ [ unwords [ nt, ":", rule l ] ]
+        , map (("  | " ++) . rule) ls
+        , [ "  ;" ]
+        ]
+      where
+        rule (p,a) = unwords [ p, "{", a , "}" ]
+        nt = nonterminal cat
 
 
 -- For every non-terminal, we construct a set of rules. A rule is a sequence of
@@ -211,8 +216,8 @@ generatePatterns terminal r = case rhsRule r of
      Right s -> terminal s
    metas its = [ ('$': show i) | (i, Left _c) <- zip [1 ::Int ..] its ]
 
-specialRules :: CF -> String
-specialRules cf = unlines $ (`map` literals cf) $ \case
+specialRules :: CF -> [String]
+specialRules cf = (`map` literals cf) $ \case
   "Ident"   -> "ident : TOK_Ident  { Ident $1 };"
   "String"  -> "string : TOK_String { $1 };"
   "Integer" -> "int :  TOK_Integer  { $1 };"

